rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+        function `fn` which specifies the computation, and an (optional)
+        `InputIterator` `inputs` to be fed into`fn`. It calls `fn` with one
+        element from the input iterator within distribution strategy `scope`.
+        Variables and optimizers for the model can be created inside or outside
+        `fn`, as long as they're created within `scope`. `run` can be called
+        multiple times with different `fn`. In graph mode, run simply creates
+        the replicated graph whereas in eager mode, it will run the actual
+        computation. It then returns the list of outputs of `fn` from each
+        replica.
+    *   `prepare_input`: This method is used to prepare the input for being
+        passed to run. It takes an `InputFn` `fn` which either returns a
+        `Dataset` or a function which returns
+        ([nests](https://www.tensorflow.org/api_docs/python/tf/contrib/framework/nest)
+        of) input tensors. It also takes an optional `InputReplicationMode`
+        `replication_mode` which specifies how, if at all, the input fn should
+        be replicated. The current options are (1) no replication, (2) replicate
+        per worker (default option), and (3) replicate per compute-replica. See
+        [InputReplicationMode](#inputreplicationmode) section for more details.
+        It returns an `InputIterator` which should be passed into `run` for
+        consumption according to the distribution strategy. `prepare_input` can
+        be called multiple times as well, potentially to get different inputs
+        being passed to different calls to `run`.
+    *   `initialize` / `finalize`: This returns ops that should be run before
+        and after running replicated computations, respectively. Typically this
+        would contain things like TPU initialize / finalize ops, and dataset
+        iterator initialize ops. In eager mode, this will instead run the
+        initialize and finalize ops directly instead of returning them.
+*   `ReplicaContext`: This is an object that is passed as the first argument to
+    `fn` when called with `run`. It contains information about the replication
+    context, such as the number of replicas, current replica id etc. It also
+    provides the methods for communication across replicas such as `all_sum`,
+    `broadcast`, etc.
+*   Input: Replicating and distributing the input correctly across replicas is
+    an important component of replicating the computation. The API provides a
+    number of input related functionality to support different use cases.
+    *   `InputReplicationMode`: The user may want to use the same input pipeline
+        for all their replicas, or have a separate input pipeline per worker or
+        per replica. This mode allows the user to specify which mode they want.
+        The knowledge of this mode helps us create the appropriate input
+        iterator. See [InputReplicationMode](#inputreplicationmode) for more
+        details.
```

Comment:
It seems like this is a splitting of the information that should in theory all be contained in DistributionStrategy. Why should we be passing information about how to replicate anywhere outside the DistStrat config?

Reply:
This is part of the `prepare_input` method because one may want to handle different types of input differently, even when using the same distribution strategy. For instance, you may use the same distribution strategy for training and evaluation but use different input replication mode.
However, I do see the alternative that we require users to use a different strategy in that case. I will add an open question at the end to discuss this as I can see merits for both options.